rules/prefer-error-cause.js

```diff
@@ -0,0 +1,320 @@
+'use strict';
+const {isCommaToken} = require('eslint-utils');
+const _ = require('lodash');
+const {appendArgument} = require('./fix/index.js');
+const getReferences = require('./utils/get-references.js');
+const getMatchingAncestorOfType = require('./utils/get-matching-ancestor-of-type.js');
+
+const ERROR = 'error';
+const SUGGESTION = 'suggestion';
+
+const messages = {
+	[ERROR]: 'Prefer specifying an old error as `cause` option where rethrowing the error.',
+	[SUGGESTION]: 'Specify the `cause` option in the Error\'s constructor.',
+};
+
+const throwStatementSelector = 'ThrowStatement';
+
+const peekTop = array => array[array.length - 1];
+
+const hasCauseProperty = objectExpression => {
+	if (!objectExpression || objectExpression.type !== 'ObjectExpression') {
+		return false;
+	}
+
+	return objectExpression.properties.some(property =>
+		!property.computed && property.key.name === 'cause',
+	);
+};
+
+const reportCannotBeFixed = (node, context) => {
+	context.report({
+		node,
+		messageId: ERROR,
+	});
+};
+
+const fixerUtils = {
+	insertProperty(fixer, node, text, sourceCode) {
+		const [penultimateToken, lastToken] = sourceCode.getLastTokens(node, 2);
+		if (node.properties.length > 0) {
+			text = isCommaToken(penultimateToken) ? `${text}` : `, ${text}`;
+		}
+
+		return fixer.insertTextBefore(lastToken, text);
+	},
+
+	insertErrorCatchClauseParameter({fixer, catchBlock, errorArgumentIdentifier}) {
+		return fixer.insertTextBefore(catchBlock.body, `(${errorArgumentIdentifier}) `);
+	},
+
+	insertFunctionParameter({fixer, catchBlock, sourceCode, errorArgumentIdentifier}) {
+		const functionExpression = catchBlock.arguments[0];
+
+		let target;
+		if (functionExpression.type === 'FunctionExpression') {
+			target = sourceCode.getTokenBefore(functionExpression.body);
+		} else if (functionExpression.type === 'ArrowFunctionExpression') {
+			target = sourceCode.getTokenBefore(sourceCode.getTokenBefore(functionExpression.body));
+		}
+
+		return fixer.insertTextBefore(target, errorArgumentIdentifier);
+	},
+};
+
+const fix = ({
+	fixer,
+	catchBlock,
+	context,
+	statementToFix,
+	errorArgumentIdentifier,
+	errorConstructorLastArgument,
+}) => {
+	const sourceCode = context.getSourceCode();
+
+	if (!errorArgumentIdentifier) {
+		const allReferences = getReferences(context.getScope());
+
+		// Cannot be fixed
+		if (allReferences.some(reference => reference?.identifier.name === 'error')) {
+			return;
+		}
+
+		errorArgumentIdentifier = 'error';
+
+		// In case of Promise#catch
+		if (catchBlock.type === 'CallExpression') {
+			return fixerUtils.insertFunctionParameter({fixer, catchBlock, sourceCode, errorArgumentIdentifier});
+		}
+
+		return fixerUtils.insertErrorCatchClauseParameter({fixer, catchBlock, errorArgumentIdentifier});
```

Comment:
Why don't we put this one line fix here directly, this just making the code hard to read.